[16.0][IMP] stock_picking_group_by_partner_by_carrier: Control the original_group_id propagation.

lmignon opened the pull request.

lmignon committed:
[FIX] stock_picking_group_by_partner_by_carrier: Don't propagate merged group

When merging an outgoing picking, update only the current moves with the new procurement group. This is required to ensure that an update of the quantity on a SO will be properly applied to all the picking chain through the run of stock rules.

lmignon committed:
[FIX] stock_picking_group_by_partner_by_carrier: Makes original group propagation configurable on stock rule.

In a multi steps delivery process, when you allow to group the moves linked to the same carrier and partner into the same shipping, it's important to be able to merge internal moves that could be pulled. This merge is only possible if the original_group_id is not propagated to the pulled moves. We add an option to disable the propagation on the stock.rule.

lmignon commented:
cc @rousseldenis @jbaudoux

rousseldenis commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -27,6 +27,12 @@ def _prepare_merge_moves_distinct_fields(self):
+    def _keep_original_group(self):
+        """Keep the original group on the move when creating a joint group."""
+        for move in self:
+            if move.group_id and not move.original_group_id:
```
Nothing more about printed pickings and state ?

rousseldenis reviewed (approved):
Code review

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_rule.py:
```diff
@@ -2,12 +2,26 @@
+        help="Propagate the original group on the created moves. This allows to "
+        "prevent merging pulled moves. This allows to cancel a SO without "
+        "canceling pulled moves from other SO (as we ensure they are not "
+        "merged). If your warehouse is configured to make the delivery in two "
+        "steps, you should not propagate the original group on the picking if your"
+        "pick picking type is also configured to group pickings. In this way, "
+        "the generated moves for the same product but for different sales orders "
+        "will be merged into one unique move. If you propagate the original group "
+        "on the picking, the moves will not be merged and you will have one move "
+        "per sale order.",
```
I would not talk about configuring merging on non outgoing moves as this should not be possible. The option is only visible on outgoing picking types. We should add a constrain to make it more explicit.
```suggestion
        "merged). You likely need this option if you don't propagate the merged procurement group to the pulled moves.",
```

jbaudoux commented on stock_picking_group_by_partner_by_carrier/tests/test_grouping.py:
```diff
@@ -267,6 +277,73 @@ def test_delivery_multi_step_group_pick(self):
         self.assertEqual(len(picks), 1)
         self.assertFalse(so1.picking_ids - so2.picking_ids)
 
+    def test_delivery_multi_step_group_out_pick_op_merged(self):
```
This should not be possible

jbaudoux replied:
I would drop this test that does not make sense as you should not be able to enable grouping on non outgoing picking type

jbaudoux reviewed (changes requested):
This overall looks good to me

lmignon commented:
Back to draft. Compatibility issue with the stock_available_to_promise_release.

lmignon committed:
[IMP] stock_picking_group_by_partner_by_carrier: Avoid procument group swap

When the grouping functionality is enabled. A new procurement group is created the first time a pick out is created as merged. Every time moves from a new group are added into the picking, the merged procurement is updated to link the new group. In this way, the group is never updated after its original creation

lmignon committed:
[IMP] stock_picking_group_by_partner_by_carrier: Initialize is_merged

lmignon committed:
[FIX] stock_picking_group_by_partner_by_carrier: backorders mgmt

When a backorder is created from a merged picking, a new procurement group must be created to ensure that only original sale orders and procurement group from the remaining moves are linked to the backorder.

jbaudoux commented:
@lmignon Can you drop changes to generated files?

lmignon committed:
[IMP] stock_picking_group_by_partner_by_carrier: Remove useless test

lmignon commented:
@jbaudoux tests removed and rebase done
